libsolidity/parsing/Parser.cpp

```diff
@@ -886,6 +886,20 @@ ASTPointer<VariableDeclaration> Parser::parseVariableDeclaration(
 					}
 				}
 			}
+			else if (
```

Comment:
I think this should have an additional condition of either
```
(_options.kind == VarDeclKind::State || _options.allowLocationSpecifier) &&
```
in case we already want to parse this as data location outside state variables - or
```
_options.kind == VarDeclKind::State &&
```
if we want to parse only state variables for now.

An example currently that neither satisfies ``_options.kind == VarDeclKind::State`` nor `` _options.allowLocationSpecifier`` is event arguments - as indicated by 
[test/libsolidity/syntaxTests/events/transient_indexed_parameter.sol](https://github.com/ethereum/solidity/pull/15001/files#diff-28e8d317223f44759ea678044aac2527515f91384ccb76f4fcf55b83d78163e2), which should probably fail to parse.
I'm actually not quite sure why that test doesn't fail the antlr grammar tests - I think it should - the grammar for ``eventParameter`` doesn't allow data locations either.

Reply:
Respectively, the cleaner way to do it would probably be to *only* do it if ``_options.kind == VarDeclKind::State &&`` here - and control parsing all other cases by whether or not we change ``TokenTraits::isLocationSpecifier(token)``

Reply:
@ekpyron , after discussing this with @cameel , we decided to restrict to only state variables.